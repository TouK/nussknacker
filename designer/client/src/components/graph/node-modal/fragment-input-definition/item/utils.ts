import { FragmentInputParameter } from ".";
import { ReturnedType } from "../../../../../types";
import { v4 as uuid4 } from "uuid";

export const getDefaultFields = (refClazzName: string): FragmentInputParameter => {
    return {
        uuid: uuid4(),
        name: "",
        required: false,
        hintText: "",
        initialValue: null,
        // fixedValuesType: FixedValuesType.UserDefinedList,
        valueEditor: null,
        // fixedValuesListPresetId: "",
        // presetSelection: "",
        valueCompileTimeValidation: null,
        typ: { refClazzName } as ReturnedType,
    };
};
