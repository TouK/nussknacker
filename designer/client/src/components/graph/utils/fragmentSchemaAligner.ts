import fp from "lodash/fp";
import { NodeType, ProcessDefinitionData } from "../../../types";

/*
 * TODO: It's a workaround
 * There may be situations where an open scenario has its fragment parameters edited by a different owner
 * In such cases, we need to align the fragmentInput properties with the updated state of the fragment process definition data.
 */
export function alignFragmentWithSchema(processDefinitionData: ProcessDefinitionData, fragmentNode: NodeType) {
    const fragmentId = fragmentNode.ref.id;
    const fragmentSchema = processDefinitionData.componentGroups
        // It's a workaround, We cannot looking for fragment schema only in the fragments group,
        // because, fragmentInput can be moved to the different group
        // and in this case, there was an error, that's why we need to iterate all groups
        .flatMap((componentGroups) => componentGroups.components)
        .find((obj) => obj?.node?.ref?.id === fragmentId);
    const fragmentSchemaParameters = fragmentSchema.node.ref.parameters;
    const mergedParameters = fragmentSchemaParameters.map(
        (param) => fragmentNode.ref.parameters.find((nodeParam) => nodeParam.name === param.name) || param,
    );
    return fp.set("ref.parameters", mergedParameters, fragmentNode);
}
