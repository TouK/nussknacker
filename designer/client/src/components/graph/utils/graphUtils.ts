import NodeUtils from "../NodeUtils";
import { cloneDeep, isEqual, map, reject } from "lodash";
import { Edge, NodeId, NodeType, ScenarioGraph, ProcessDefinitionData } from "../../../types";
import {
    adjustBranchParametersAfterDisconnect,
    enrichNodeWithProcessDependentData,
    removeBranchParameter,
} from "../../../reducers/graph/utils";
import { dia } from "jointjs";
import { isTouchEvent } from "../../../helpers/detectDevice";

export function mapProcessWithNewNode(scenarioGraph: ScenarioGraph, before: NodeType, after: NodeType): ScenarioGraph {
    return {
        ...scenarioGraph,
        edges: map(scenarioGraph.edges, (e) => {
            if (isEqual(e.from, before.id)) {
                return { ...e, from: after.id, to: e.to };
            } else if (isEqual(e.to, before.id)) {
                return { ...e, from: e.from, to: after.id };
            } else {
                return e;
            }
        }),
        nodes: map(scenarioGraph.nodes, (n) => {
            return isEqual(n, before) ? after : mapBranchParametersWithNewNode(before.id, after.id, n);
        }),
    };
}

//we do mapping here, because we validate changed process before closing modal, and before applying state change in reducer.
function mapBranchParametersWithNewNode(beforeId: NodeId, afterId: NodeId, node: NodeType): NodeType {
    if (beforeId !== afterId && node.branchParameters?.find((bp) => bp.branchId === beforeId)) {
        const newNode = cloneDeep(node);
        const branchParameter = newNode.branchParameters.find((bp) => bp.branchId === beforeId);
        if (branchParameter) {
            branchParameter.branchId = afterId;
        }
        return newNode;
    } else {
        return node;
    }
}

export function mapProcessWithNewEdge(scenarioGraph: ScenarioGraph, before: Edge, after: Edge): ScenarioGraph {
    return {
        ...scenarioGraph,
        edges: map(scenarioGraph.edges, (e) => {
            if (isEqual(e.from, before.from) && isEqual(e.to, before.to)) {
                return after;
            } else {
                return e;
            }
        }),
    };
}

export function replaceNodeOutputEdges(
    scenarioGraph: ScenarioGraph,
    processDefinitionData: ProcessDefinitionData,
    edgesAfter: Edge[],
    nodeId: NodeType["id"],
): ScenarioGraph {
    const edgesBefore = scenarioGraph.edges.filter(({ from }) => from === nodeId);

    if (isEqual(edgesBefore, edgesAfter)) return scenarioGraph;

    const oldTargets = new Set(edgesBefore.map((e) => e.to));
    const newTargets = new Set(edgesAfter.map((e) => e.to));
    return {
        ...scenarioGraph,
        edges: scenarioGraph.edges.filter(({ from }) => !(from === nodeId)).concat(edgesAfter),
        nodes: scenarioGraph.nodes.map((node) => {
            if (newTargets.has(node.id)) {
                return enrichNodeWithProcessDependentData(node, processDefinitionData, edgesAfter);
            }
            if (oldTargets.has(node.id)) {
                return removeBranchParameter(node, nodeId);
            }
            return node;
        }),
    };
}

export function deleteNode(scenarioGraph: ScenarioGraph, nodeId: NodeId): ScenarioGraph {
    const edges = scenarioGraph.edges.filter((e) => e.from !== nodeId).map((e) => (e.to === nodeId ? { ...e, to: "" } : e));
    const nodesAfterRemove = scenarioGraph.nodes.filter((n) => n.id !== nodeId);
    const removedEdges = scenarioGraph.edges.filter((e) => e.from === nodeId);
    const nodes = adjustBranchParametersAfterDisconnect(nodesAfterRemove, removedEdges);

    return { ...scenarioGraph, edges, nodes };
}

export function canInjectNode(
    scenarioGraph: ScenarioGraph,
    sourceId: NodeId,
    middleManId: NodeId,
    targetId: NodeId,
    processDefinitionData: ProcessDefinitionData,
): boolean {
    const processAfterDisconnection = deleteEdge(scenarioGraph, sourceId, targetId);
    const canConnectSourceToMiddleMan = NodeUtils.canMakeLink(sourceId, middleManId, processAfterDisconnection, processDefinitionData);
    const processWithConnectedSourceAndMiddleMan = addEdge(processAfterDisconnection, sourceId, middleManId);
    const canConnectMiddleManToTarget = NodeUtils.canMakeLink(
        middleManId,
        targetId,
        processWithConnectedSourceAndMiddleMan,
        processDefinitionData,
    );
    return canConnectSourceToMiddleMan || canConnectMiddleManToTarget;
}

function deleteEdge(scenarioGraph: ScenarioGraph, fromId: NodeId, toId: NodeId): ScenarioGraph {
    return {
        ...scenarioGraph,
        edges: reject(scenarioGraph.edges, (e) => e.from === fromId && e.to === toId),
    };
}

function addEdge(scenarioGraph: ScenarioGraph, fromId: NodeId, toId: NodeId): ScenarioGraph {
    return {
        ...scenarioGraph,
        edges: scenarioGraph.edges.concat({ from: fromId, to: toId }),
    };
}

export const handleGraphEvent =
    <T = dia.CellView>(
        touchEvent: ((view: T, event: dia.Event) => void) | null,
        mouseEvent: ((view: T, event: dia.Event) => void) | null,
    ) =>
    (view: T, event: dia.Event) =>
        isTouchEvent(event) ? touchEvent?.(view, event) : mouseEvent?.(view, event);
