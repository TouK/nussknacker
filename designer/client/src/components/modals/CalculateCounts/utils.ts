import { TFunction } from "i18next";
import { Range } from "./CountsRangesButtons";
import moment, { Moment } from "moment";

export function predefinedRanges(t: TFunction<string>): Range[] {
    const forDay = (name: string, moment: () => Moment): Range => ({
        name: name,
        from: () => moment().startOf("day"),
        to: () => moment().endOf("day"),
    });

    return [
        {
            name: t("calculateCounts.range.lastHour", "Last hour"),
            from: () => moment().subtract(1, "hour"),
            to: () => moment(),
        },
        forDay(t("calculateCounts.range.today", "Today"), () => moment()),
        forDay(t("calculateCounts.range.yesterday", "Yesterday"), () => moment().subtract(1, "day")),
        forDay(t("calculateCounts.range.dayBeforeYesterday", "Day before yesterday"), () => moment().subtract(2, "day")),
        forDay(t("calculateCounts.range.thisDayLastWeek", "This day last week"), () => moment().subtract(7, "day")),
    ];
}
