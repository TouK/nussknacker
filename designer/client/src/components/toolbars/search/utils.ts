import { Edge, NodeType } from "../../../types";
import { isEqual, uniq } from "lodash";
import { useSelector } from "react-redux";
import { getScenario } from "../../../reducers/selectors/graph";
import NodeUtils from "../../graph/NodeUtils";
import { useMemo } from "react";
import { useTranslation } from "react-i18next";
import { ensureArray } from "../../../common/arrayUtils";
import { SearchQuery } from "./SearchResults";
import { getComponentGroups } from "../../../reducers/selectors/settings";

type SelectorResult = { expression: string } | string;
type Selector = (node: NodeType) => SelectorResult | SelectorResult[];
type FilterSelector = { name: string; selector: Selector }[];

const fieldsSelectors: FilterSelector = [
    {
        name: "name",

        selector: (node) => node.id,
    },
    {
        name: "description",
        selector: (node) => node.additionalFields?.description,
    },
    {
        name: "type",
        selector: (node) => [node.nodeType, node.ref?.typ, node.ref?.id, node.type, node.service?.id],
    },
    {
        name: "value",
        selector: (node) => node.ref?.outputVariableNames && Object.values(node.ref?.outputVariableNames),
    },
    {
        name: "label",
        selector: (node) => node.ref?.outputVariableNames && Object.keys(node.ref?.outputVariableNames),
    },
    {
        name: "value",
        selector: (node) => [node.expression, node.exprVal, node.value],
    },
    {
        name: "output",
        selector: (node) => [node.outputName, node.output, node.outputVar, node.varName],
    },
    {
        name: "value",
        selector: (node) => [node.parameters, node.ref?.parameters, node.service?.parameters, node.fields].flat().map((p) => p?.expression),
    },
    {
        name: "label",
        selector: (node) => [node.parameters, node.ref?.parameters, node.service?.parameters, node.fields].flat().map((p) => p?.name),
    },
];

function matchFilters(value: SelectorResult, filterValues: string[]): boolean {
    const resolved = typeof value === "string" ? value : value?.expression;
    return filterValues.length && filterValues.some((filter) => filter != "" && resolved?.toLowerCase().includes(filter.toLowerCase()));
}

export const findFields = (filterValues: string[], node: NodeType) => {
    return uniq(
        fieldsSelectors.flatMap(({ name, selector }) =>
            ensureArray(selector(node))
                .filter((v) => matchFilters(v, filterValues))
                .map(() => name),
        ),
    );
};

const findFieldsUsingSelectorWithName = (selectorName: string, filterValues: string[], node: NodeType) => {
    return uniq(
        fieldsSelectors
            .filter((selector) => selector.name == selectorName)
            .flatMap(({ name, selector }) =>
                ensureArray(selector(node))
                    .filter((v) => matchFilters(v, filterValues))
                    .map(() => name),
            ),
    );
};

function useComponentTypes(): Set<string> {
    const componentsGroups = useSelector(getComponentGroups);

    return useMemo(() => {
        return new Set(
            componentsGroups.flatMap((componentGroup) => componentGroup.components).map((component) => component.label.toLowerCase()),
        );
    }, []);
}

export function useNodeTypes(): string[] {
    const { scenarioGraph } = useSelector(getScenario);
    const allNodes = NodeUtils.nodesFromScenarioGraph(scenarioGraph);
    const componentsSet = useComponentTypes();

    return useMemo(() => {
        const nodeSelector = fieldsSelectors.find((selector) => selector.name == "type")?.selector;
        const availableTypes = allNodes
            .flatMap((node) => ensureArray(nodeSelector(node)).filter((item) => item !== undefined))
            .map((selectorResult) => (typeof selectorResult === "string" ? selectorResult : selectorResult?.expression))
            .filter((type) => componentsSet.has(type.toLowerCase()));

        return uniq(availableTypes);
    }, [allNodes]);
}

export function useFilteredNodes(searchQuery: SearchQuery): {
    groups: string[];
    node: NodeType;
    edges: Edge[];
}[] {
    const { t } = useTranslation();
    const { scenarioGraph } = useSelector(getScenario);
    const allNodes = NodeUtils.nodesFromScenarioGraph(scenarioGraph);
    const allEdges = NodeUtils.edgesFromScenarioGraph(scenarioGraph);

    const searchKeys = Object.keys(searchQuery).filter((key) => searchQuery[key as keyof SearchQuery] !== undefined);
    const isSimpleSearch = searchKeys.length === 1 && searchKeys[0] === "plainQuery";

    const displayNames = useMemo(
        () => ({
            name: t("panels.search.field.id", "Name"),
            description: t("panels.search.field.description", "Description"),
            label: t("panels.search.field.label", "Label"),
            value: t("panels.search.field.value", "Value"),
            output: t("panels.search.field.output", "Output"),
            edge: t("panels.search.field.edge", "Edge"),
            type: t("panels.search.field.type", "Type"),
        }),
        [t],
    );

    return useMemo(
        () =>
            allNodes
                .map((node) => {
                    let edges: Edge[] = [];
                    let groups: string[] = [];

                    if (isSimpleSearch) {
                        edges = allEdges
                            .filter((e) => e.from === node.id)
                            .filter((e) => matchFilters(e.edgeType?.condition, [searchQuery.plainQuery]));

                        groups = findFields([searchQuery.plainQuery], node)
                            .concat(edges.length ? "edge" : null)
                            .map((name) => displayNames[name])
                            .filter(Boolean);

                        return { node, edges, groups };
                    } else {
                        const edgesAux = allEdges
                            .filter((e) => e.from === node.id)
                            .filter((e) => matchFilters(e.edgeType?.condition, [searchQuery.plainQuery]));

                        const groupsAux: string[] = findFields([searchQuery.plainQuery], node)
                            .concat(edgesAux.length ? "edge" : null)
                            .map((name) => displayNames[name])
                            .filter(Boolean);

                        edges =
                            "edge" in searchQuery
                                ? allEdges
                                      .filter((e) => e.from === node.id)
                                      .filter((e) => matchFilters(e.edgeType?.condition, searchQuery.edge))
                                : [];

                        const keyNamesRelevantForFiltering = Object.keys(searchQuery).filter(
                            (key) => key !== "searchType" && key !== "plainQuery",
                        );

                        const displayKeyNamesRelevantForFiltering: string[] = keyNamesRelevantForFiltering.map(
                            (name) => displayNames[name],
                        );

                        groups = keyNamesRelevantForFiltering
                            .map((key) => findFieldsUsingSelectorWithName(key, searchQuery[key], node))
                            .flat()
                            .concat(edges.length ? "edge" : null)
                            .map((name) => displayNames[name])
                            .filter(Boolean);

                        const allChecksPassed = isEqual(groups, displayKeyNamesRelevantForFiltering);

                        const hasValidQueryOrGroups = searchQuery.plainQuery === "" ? false : groupsAux.length === 0;

                        if (!allChecksPassed || hasValidQueryOrGroups) {
                            edges = [];
                            groups = [];
                        }

                        return { node, edges, groups };
                    }
                })
                .filter(({ groups }) => groups.length),
        [displayNames, allEdges, searchQuery, allNodes],
    );
}

export function resolveSearchQuery(text: string): SearchQuery {
    const result: SearchQuery = {};
    const regex = /(\w+):\(([^)]*)\)/g;
    let match: RegExpExecArray | null;
    let lastIndex = 0;

    while ((match = regex.exec(text)) !== null) {
        const key = match[1] as keyof Omit<SearchQuery, "plainQuery">;
        const values = match[2].split(",").map((value) => value.trim().replace(/"/g, ""));
        result[key] = values.length > 0 ? values : [];
        lastIndex = regex.lastIndex;
    }

    result.plainQuery = text.slice(lastIndex).trim();

    return result;
}

export function searchQueryToString(query: SearchQuery): string {
    const plainQuery = query.plainQuery;

    const formattedParts = Object.entries(query)
        .filter(([key]) => key !== "plainQuery")
        .map(([key, value]) => {
            if (Array.isArray(value) && !(value.length === 1 && value[0] === "")) {
                return `${key}:(${value})`;
            } else if (typeof value === "string" && value.length > 0) {
                return `${key}:(${[value]})`;
            }
            return []; // Skip undefined or invalid values
        })
        .filter(Boolean) // Remove null values
        .join(" "); // Join the formatted parts with a space

    // Append plainQuery at the end without a key if it exists
    return plainQuery
        ? `${formattedParts} ${plainQuery}`.trim() // Ensure no trailing space
        : formattedParts;
}
