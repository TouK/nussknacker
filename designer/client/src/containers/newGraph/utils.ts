import React, { useContext, useEffect } from "react";

export function createContextHook<T>(Context: React.Context<T>, Provider: React.ComponentType): () => T;
export function createContextHook<T, C>(Context: React.Context<T>, Provider: React.ComponentType, extendFn: (ctx: T) => C): () => C;
export function createContextHook<T, C>(Context: React.Context<T>, Provider: React.ComponentType, extendFn?: (ctx: T) => C) {
    return () => {
        const context = useContext(Context);

        if (!context) throw new Error(`used outside ${Provider.name}`);

        if (typeof extendFn === "function") {
            return extendFn(context);
        }

        return context;
    };
}

export function useContextForward<C>(forwardedRef: React.ForwardedRef<C>, context: C) {
    useEffect(() => {
        if (typeof forwardedRef === "function") {
            forwardedRef(context);
        } else if (forwardedRef) {
            forwardedRef.current = context;
        }
    }, [context, forwardedRef]);
}
