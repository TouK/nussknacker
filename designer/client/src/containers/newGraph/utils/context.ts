import React, { useContext, useEffect, useMemo } from "react";

const getComponentName = (Component: React.ComponentType): string => {
    if (Component.displayName) return Component.displayName;
    return (Component["render"] || Component).name;
};

export function createContextHook<T>(Context: React.Context<T>, Provider: React.ComponentType): () => T;
export function createContextHook<T, C>(Context: React.Context<T>, Provider: React.ComponentType, extendFn: (ctx: T) => C): () => C;
export function createContextHook<T, C>(Context: React.Context<T>, Provider: React.ComponentType, extendFn?: (ctx: T) => C) {
    return function () {
        const context = useContext(Context);

        return useMemo(() => {
            if (!context) {
                throw new Error(`context hook used outside ${getComponentName(Provider) || "context provider"}`);
            }
            if (typeof extendFn === "function") {
                return extendFn(context);
            }
            return context;
        }, [context]);
    };
}

export function useContextForward<C>(forwardedRef: React.ForwardedRef<C>, context: C) {
    useEffect(() => {
        if (typeof forwardedRef === "function") {
            forwardedRef(context);
        } else if (forwardedRef) {
            forwardedRef.current = context;
        }
    }, [context, forwardedRef]);
}
