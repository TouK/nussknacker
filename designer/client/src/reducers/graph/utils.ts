import { cloneDeep, Dictionary, map, reject, zipObject, zipWith } from "lodash";
import { Layout, NodePosition, NodesWithPositions } from "../../actions/nk";
import ProcessUtils from "../../common/ProcessUtils";
import { ExpressionLang } from "../../components/graph/node-modal/editors/expression/types";
import NodeUtils from "../../components/graph/NodeUtils";
import { BranchParams, Edge, EdgeType, NodeId, NodeType, ProcessDefinitionData } from "../../types";
import { GraphState } from "./types";
import { StickyNote } from "../../common/StickyNote";
import { createStickyNoteId } from "../../types/stickyNote";

export function updateLayoutAfterNodeIdChange(layout: Layout, oldId: NodeId, newId: NodeId): Layout {
    return map(layout, (n) => (oldId === n.id ? { ...n, id: newId } : n));
}

export function updateAfterNodeDelete(state: GraphState, idToDelete: NodeId) {
    return {
        ...state,
        scenario: {
            ...state.scenario,
            scenarioGraph: {
                ...state.scenario.scenarioGraph,
                nodes: state.scenario.scenarioGraph.nodes.filter((n) => n.id !== idToDelete),
            },
        },
        layout: state.layout.filter((n) => n.id !== idToDelete),
    };
}

function generateUniqueNodeId(initialId: NodeId, usedIds: NodeId[], nodeCounter: number, isCopy: boolean): NodeId {
    const newId = isCopy ? `${initialId} (copy ${nodeCounter})` : `${initialId} ${nodeCounter}`;
    return usedIds.includes(newId) ? generateUniqueNodeId(initialId, usedIds, nodeCounter + 1, isCopy) : newId;
}

function createUniqueNodeId(initialId: NodeId, usedIds: NodeId[], isCopy: boolean): NodeId {
    return initialId && !usedIds.includes(initialId) ? initialId : generateUniqueNodeId(initialId, usedIds, 1, isCopy);
}

function getUniqueIds(initialIds: string[], alreadyUsedIds: string[], isCopy: boolean) {
    return initialIds.reduce((uniqueIds, initialId) => {
        const reservedIds = alreadyUsedIds.concat(uniqueIds);
        const uniqueId = createUniqueNodeId(initialId, reservedIds, isCopy);
        return uniqueIds.concat(uniqueId);
    }, []);
}

function adjustBranchParameters(branchParameters: BranchParams[], uniqueIds: string[]) {
    return branchParameters?.map(({ branchId, ...branchParameter }: BranchParams) => ({
        ...branchParameter,
        branchId: uniqueIds.find((uniqueId) => uniqueId.includes(branchId)),
    }));
}

export function prepareNewNodesWithLayout(
    currentNodes: NodeType[] = [],
    newNodesWithPositions: NodesWithPositions,
    isCopy: boolean,
): {
    layout: NodePosition[];
    nodes: NodeType[];
    idMapping: Dictionary<string>;
} {
    const newNodes = newNodesWithPositions.map(({ node }) => node);
    const newPositions = newNodesWithPositions.map(({ position }) => position);
    const alreadyUsedIds = currentNodes.map((node) => node.id);
    const initialIds = newNodes.map(({ id }) => id);
    const uniqueIds = getUniqueIds(initialIds, alreadyUsedIds, isCopy);

    return {
        nodes: zipWith(newNodes, uniqueIds, (node, id) => ({
            ...node,
            id,
            branchParameters: adjustBranchParameters(node.branchParameters, uniqueIds),
        })),
        layout: zipWith(newPositions, uniqueIds, (position, id) => ({
            id,
            position,
        })),
        idMapping: zipObject(initialIds, uniqueIds),
    };
}

export function removeStickyNoteFromLayout(state: GraphState, stickyNoteId: number): { layout: NodePosition[]; stickyNotes: StickyNote[] } {
    const { layout } = state;
    const updatedStickyNotes = state.stickyNotes.filter((n) => n.noteId != stickyNoteId);
    const updatedLayout = updatedStickyNotes.map((stickyNote) => {
        return { id: createStickyNoteId(stickyNote.noteId), position: stickyNote.layoutData };
    });
    return {
        stickyNotes: [...updatedStickyNotes],
        layout: [...layout, ...updatedLayout],
    };
}

export function prepareNewStickyNotesWithLayout(
    state: GraphState,
    stickyNotes: StickyNote[],
): { layout: NodePosition[]; stickyNotes: StickyNote[] } {
    const { layout } = state;
    const updatedLayout = stickyNotes.map((stickyNote) => {
        return { id: createStickyNoteId(stickyNote.noteId), position: stickyNote.layoutData };
    });
    return {
        stickyNotes: [...stickyNotes],
        layout: [...layout, ...updatedLayout],
    };
}

export function addNodesWithLayout(
    state: GraphState,
    changes: {
        nodes: NodeType[];
        layout: NodePosition[];
        edges?: Edge[];
    },
): GraphState {
    const { nodes = [], edges = [], ...scenarioGraph } = state.scenario.scenarioGraph;
    const nextNodes = [...nodes, ...changes.nodes];
    const nextEdges = changes.edges || edges;
    const nextLayout = [...state.layout, ...changes.layout];
    return {
        ...state,
        scenario: {
            ...state.scenario,
            scenarioGraph: {
                ...scenarioGraph,
                nodes: nextNodes,
                edges: nextEdges,
            },
        },
        layout: nextLayout,
    };
}


export function addStickyNotesWithLayout(
    state: GraphState,
    { stickyNotes, layout }: ReturnType<typeof prepareNewStickyNotesWithLayout>,
): GraphState {
    return {
        ...state,
        stickyNotes: stickyNotes,
        layout,
    };
}

export function createEdge(
    fromNode: NodeType,
    toNode: NodeType,
    edgeType: EdgeType,
    nodeOutputEdges: Edge[],
    processDefinitionData: ProcessDefinitionData,
) {
    const baseEdge = { from: fromNode?.id, to: toNode?.id };
    const adjustedEdgeType = edgeType || NodeUtils.getNextEdgeType(nodeOutputEdges, fromNode, processDefinitionData);
    return adjustedEdgeType ? { ...baseEdge, edgeType: adjustedEdgeType } : baseEdge;
}

export function removeBranchParameter(node: NodeType, branchId: NodeId) {
    const { branchParameters, ...clone } = cloneDeep(node);
    return {
        ...clone,
        branchParameters: reject(branchParameters, (parameter) => parameter.branchId === branchId),
    };
}

export function adjustBranchParametersAfterDisconnect(nodes: NodeType[], removedEdges: Pick<Edge, "from" | "to">[]): NodeType[] {
    return removedEdges.reduce((resultNodes, { from, to }) => {
        const node = resultNodes.find((n) => n.id === to);
        if (node && NodeUtils.nodeIsJoin(node)) {
            const newToNode = removeBranchParameter(node, from);
            return resultNodes.map((n) => {
                return n.id === to ? newToNode : n;
            });
        } else {
            return resultNodes;
        }
    }, nodes);
}

export function enrichNodeWithProcessDependentData(
    originalNode: NodeType,
    processDefinitionData: ProcessDefinitionData,
    edges: Edge[],
): NodeType {
    const node = cloneDeep(originalNode);

    switch (node.type) {
        case "Join": {
            const parameters = ProcessUtils.extractComponentDefinition(node, processDefinitionData.components)?.parameters;
            const declaredBranchParameters = parameters?.filter((p) => p.branchParam) || [];
            const incomingEdges = edges.filter((e) => e.to === node.id);
            const branchParameters = incomingEdges.map((edge) => {
                const branchId = edge.from;
                const existingBranchParams = node.branchParameters.find((p) => p.branchId === branchId);
                const parameters = declaredBranchParameters.map((branchParamDef) => {
                    const existingParamValue = existingBranchParams?.parameters?.find((p) => p.name === branchParamDef.name);
                    if (!existingParamValue) {
                        const templateParamValue = node.branchParametersTemplate?.find((p) => p.name === branchParamDef.name);
                        if (!templateParamValue) {
                            // We need to have this fallback to some template for situation when it is existing node and it has't got
                            // defined parameters filled. see note in DefinitionPreparer on backend side TODO: remove it after API refactor
                            return {
                                name: branchParamDef.name,
                                expression: {
                                    expression: `#${branchParamDef.name}`,
                                    language: ExpressionLang.SpEL,
                                },
                            };
                        }
                        return cloneDeep(templateParamValue);
                    }
                    return existingParamValue;
                });

                return { branchId, parameters };
            });

            return { ...node, branchParameters };
        }
    }

    return node;
}
