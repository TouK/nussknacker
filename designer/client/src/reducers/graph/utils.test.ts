import { prepareNewNodesWithLayout } from "./utils";
import { nodesWithPositions, state } from "./utils.fixtures";

describe("GraphUtils prepareNewNodesWithLayout", () => {
    it("should update union output expression parameter with an updated node name when new unique node ids created", () => {
        const expected = {
            layout: [
                {
                    id: "choice",
                    position: {
                        x: 180,
                        y: 540,
                    },
                },
                {
                    id: "variable 1",
                    position: {
                        x: 0,
                        y: 720,
                    },
                },
                {
                    id: "variable 2",
                    position: {
                        x: 360,
                        y: 720,
                    },
                },
                {
                    id: "union",
                    position: {
                        x: 180,
                        y: 900,
                    },
                },
                {
                    id: "variable 1 (copy 1)",
                    position: {
                        x: 350,
                        y: 859,
                    },
                },
                {
                    id: "variable 2 (copy 1)",
                    position: {
                        x: 710,
                        y: 859,
                    },
                },
                {
                    id: "union (copy 1)",
                    position: {
                        x: 530,
                        y: 1039,
                    },
                },
            ],
            nodes: [
                {
                    additionalFields: {
                        description: null,
                        layoutData: {
                            x: 180,
                            y: 540,
                        },
                    },
                    exprVal: null,
                    expression: null,
                    id: "choice",
                    type: "Switch",
                },
                {
                    additionalFields: {
                        description: null,
                        layoutData: {
                            x: 0,
                            y: 720,
                        },
                    },
                    id: "variable 1",
                    type: "Variable",
                    value: {
                        expression: "'value'",
                        language: "spel",
                    },
                    varName: "varName1",
                },
                {
                    additionalFields: {
                        description: null,
                        layoutData: {
                            x: 360,
                            y: 720,
                        },
                    },
                    id: "variable 2",
                    type: "Variable",
                    value: {
                        expression: "'value'",
                        language: "spel",
                    },
                    varName: "varName2",
                },
                {
                    additionalFields: {
                        description: null,
                        layoutData: {
                            x: 180,
                            y: 900,
                        },
                    },
                    branchParameters: [
                        {
                            branchId: "variable 1",
                            parameters: [
                                {
                                    expression: {
                                        expression: "1",
                                        language: "spel",
                                    },
                                    name: "Output expression",
                                },
                            ],
                        },
                        {
                            branchId: "variable 2",
                            parameters: [
                                {
                                    expression: {
                                        expression: "2",
                                        language: "spel",
                                    },
                                    name: "Output expression",
                                },
                            ],
                        },
                    ],
                    id: "union",
                    nodeType: "union",
                    outputVar: "outputVar",
                    parameters: [],
                    type: "Join",
                },
                {
                    additionalFields: {
                        description: null,
                        layoutData: {
                            x: 0,
                            y: 720,
                        },
                    },
                    id: "variable 1 (copy 1)",
                    type: "Variable",
                    value: {
                        expression: "'value'",
                        language: "spel",
                    },
                    varName: "varName1",
                },
                {
                    additionalFields: {
                        description: null,
                        layoutData: {
                            x: 360,
                            y: 720,
                        },
                    },
                    id: "variable 2 (copy 1)",
                    type: "Variable",
                    value: {
                        expression: "'value'",
                        language: "spel",
                    },
                    varName: "varName2",
                },
                {
                    additionalFields: {
                        description: null,
                        layoutData: {
                            x: 180,
                            y: 900,
                        },
                    },
                    branchParameters: [
                        {
                            branchId: "variable 1 (copy 1)",
                            parameters: [
                                {
                                    expression: {
                                        expression: "1",
                                        language: "spel",
                                    },
                                    name: "Output expression",
                                },
                            ],
                        },
                        {
                            branchId: "variable 2 (copy 1)",
                            parameters: [
                                {
                                    expression: {
                                        expression: "2",
                                        language: "spel",
                                    },
                                    name: "Output expression",
                                },
                            ],
                        },
                    ],
                    id: "union (copy 1)",
                    nodeType: "union",
                    outputVar: "outputVar",
                    parameters: [],
                    type: "Join",
                },
            ],
            uniqueIds: ["variable 1 (copy 1)", "variable 2 (copy 1)", "union (copy 1)"],
        };

        expect(prepareNewNodesWithLayout(state, nodesWithPositions, true)).toEqual(expected);
    });
});
