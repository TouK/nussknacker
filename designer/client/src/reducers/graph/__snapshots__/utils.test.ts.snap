// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphUtils prepareNewNodesWithLayout should update union output expression parameter with an updated node name when new unique node ids created 1`] = `
{
  "idMapping": {
    "union": "union (copy 1)",
    "variable 1": "variable 1 (copy 1)",
    "variable 2": "variable 2 (copy 1)",
  },
  "layout": [
    {
      "id": "variable 1 (copy 1)",
      "position": {
        "x": 350,
        "y": 859,
      },
    },
    {
      "id": "variable 2 (copy 1)",
      "position": {
        "x": 710,
        "y": 859,
      },
    },
    {
      "id": "union (copy 1)",
      "position": {
        "x": 530,
        "y": 1039,
      },
    },
  ],
  "nodes": [
    {
      "additionalFields": {
        "description": null,
        "layoutData": {
          "x": 0,
          "y": 720,
        },
      },
      "branchParameters": undefined,
      "id": "variable 1 (copy 1)",
      "label": "variable",
      "type": "Variable",
      "value": {
        "expression": "'value'",
        "language": "spel",
      },
      "varName": "varName1",
    },
    {
      "additionalFields": {
        "description": null,
        "layoutData": {
          "x": 360,
          "y": 720,
        },
      },
      "branchParameters": undefined,
      "id": "variable 2 (copy 1)",
      "label": "variable",
      "type": "Variable",
      "value": {
        "expression": "'value'",
        "language": "spel",
      },
      "varName": "varName2",
    },
    {
      "additionalFields": {
        "description": null,
        "layoutData": {
          "x": 180,
          "y": 900,
        },
      },
      "branchParameters": [
        {
          "branchId": "variable 1 (copy 1)",
          "parameters": [
            {
              "expression": {
                "expression": "1",
                "language": "spel",
              },
              "name": "Output expression",
            },
          ],
        },
        {
          "branchId": "variable 2 (copy 1)",
          "parameters": [
            {
              "expression": {
                "expression": "2",
                "language": "spel",
              },
              "name": "Output expression",
            },
          ],
        },
      ],
      "id": "union (copy 1)",
      "label": "union",
      "nodeType": "union",
      "outputVar": "outputVar",
      "parameters": [],
      "type": "Join",
    },
  ],
}
`;
